Snowflake.Data.Tests/UnitTests/SFS3ClientTest.cs

```diff
@@ -0,0 +1,273 @@
+                    stageCredentials = _stageCredentials,
+                    storageAccount = null
+                }
+            };
+
+            // Setup mock S3 client
+            MockS3Client mockClient = SetUpMockAwsClient(_fileMetadata);
+            _client = new SFS3Client(_fileMetadata.stageInfo, MaxRetry, Parallel, _proxyCredentials, mockClient); 
+            _fileMetadata.client = _client;
+
+            _cancellationToken = new CancellationToken();
+        }
+
+        internal static MockS3Client SetUpMockAwsClient(SFFileMetadata fileMetadata)
+        {
+            AmazonS3Config clientConfig = new AmazonS3Config();
+            RegionEndpoint regionEndpoint = RegionEndpoint.GetBySystemName(fileMetadata.stageInfo.region);
+            clientConfig.RegionEndpoint = regionEndpoint;
+            MockS3Client mockClient = new MockS3Client(AwsKeyId,
+                AwsSecretKey,
+                AwsToken,
+                clientConfig);
+
+            return mockClient;
+        }
+
+        [Test]
+        [Ignore("S3ClientTest")]
+        public void S3ClientTestDone()
+        {
+            // Do nothing;
+        }
+
+        [Test]
+        public void TestExtractBucketNameAndPath()
+        {
+            RemoteLocation location = _client.ExtractBucketNameAndPath(_fileMetadata.stageInfo.location);
+
+            // Split LOCATION based on the first '/' character
+            string[] bucketAndKey = Location.Split(new[] { '/' }, 2);
+
+            Assert.AreEqual(bucketAndKey[0], location.bucket);
+            Assert.AreEqual(bucketAndKey[1], location.key);
+        }
+
+        [Test]
+        [TestCase(MockS3Client.AwsStatusOk, ResultStatus.UPLOADED)]
+        [TestCase(SFS3Client.EXPIRED_TOKEN, ResultStatus.RENEW_TOKEN)]
+        [TestCase(SFS3Client.NO_SUCH_KEY, ResultStatus.NOT_FOUND_FILE)]
+        [TestCase(MockS3Client.AwsStatusError, ResultStatus.ERROR)] // Any error that isn't the above will return ResultStatus.ERROR
+        public void TestGetFileHeader(string requestKey, ResultStatus expectedResultStatus)
+        {
+            // Setup request
+            _fileMetadata.stageInfo.location = requestKey + "/" + HttpMethod.Head;
+
+            FileHeader fileHeader = _client.GetFileHeader(_fileMetadata);
+
+            AssertForGetFileHeaderTests(expectedResultStatus, fileHeader);
+        }
+
+        [Test]
+        [TestCase(MockS3Client.AwsStatusOk, ResultStatus.UPLOADED)]
+        [TestCase(SFS3Client.EXPIRED_TOKEN, ResultStatus.RENEW_TOKEN)]
+        [TestCase(SFS3Client.NO_SUCH_KEY, ResultStatus.NOT_FOUND_FILE)]
+        [TestCase(MockS3Client.AwsStatusError, ResultStatus.ERROR)] // Any error that isn't the above will return ResultStatus.ERROR
+        public async Task TestGetFileHeaderAsync(string requestKey, ResultStatus expectedResultStatus)
+        {
+            // Setup request
+            _fileMetadata.stageInfo.location = requestKey + "/" + HttpMethod.Head + "/async";
```

Comment:
let's follow arrange, act, assert pattern

Reply:
Changed the tests to follow that pattern